Refactor compiler specializations to consider backend

giuseros opened the pull request:
In this PR I am trying to refactor the specializations that we apply to the signature of a given function in Triton. 

Basically, given a kernel there are some argument properties that can help compilation. E.g., divisibility by 16 and the fact that an integer is equal to 1. 

In a previous PR: https://github.com/triton-lang/triton/pull/4716, I needed other specializations to add buffer support in the AMD backend (and get back some performance when we were using unaligned masked loads). 

So this is my attempt to redesign the specialization support to introduce per-backend specializations. The idea is that `AttrsDescriptor` is now the class that is taking care of doing the analysis of the parameters and adding the specialization. It also has a function table where more specializations can be added per-backend.

giuseros commented:
Hi @ThomasRaoux , @antiagainst , this is my best attempt to redesign the specialization collection. Please, let me know what you think

ThomasRaoux commented on python/triton/tools/compile.py:
```diff
@@ -106,9 +106,10 @@ def constexpr(s):
-    divisible_by_16 = [i for i, h in hints.items() if h == 16]
-    equal_to_1 = [i for i, h in hints.items() if h == 1]
-    attrs = triton.compiler.AttrsDescriptor(divisible_by_16=divisible_by_16, equal_to_1=equal_to_1)
+    divisible_by_16 = [(i, 16) for i, h in hints.items() if h == 16]
+    equal_to_1 = [(i, 1) for i, h in hints.items() if h == 1]
```
Why do we need to pass the extra `16` and `1` here?

giuseros replied:
Those would be the values of the specializations. Basically the value that the property of the attribute will be set to. The point is that the class is trying to do two things
1. It is taking params and arguments, and it is coming up with something like `prop1: ( (p0,val), (p3,val), (p4,val))`
2. It is combing those params together to form a list of param properties `p0 : {prop1:val, prop2:val,} ....
This is why I am passing those values around

ThomasRaoux commented on python/triton/runtime/jit.py:
```diff
@@ -467,41 +466,10 @@ def _key_of(arg):
-        return AttrsDescriptor(tuple(divisible_by_16), tuple(equal_to_1))
-        # return _triton.code_gen.instance_descriptor(divisible_by_16,
-        # equal_to_1)
+        target = driver.active.get_current_target()
```
It would be nicer if it could come directly from the backend class instead of checking the driver once again here.

giuseros replied:
So ideally, the backend class returns a list of specializations? And we can pass this to the `AttrsDescriptor`?

ThomasRaoux commented on python/triton/compiler/compiler.py:
```diff
@@ -9,32 +9,95 @@
 
-    def to_dict(self):
-        return {'divisible_by_16': list(self.divisible_by_16), 'equal_to_1': list(self.equal_to_1)}
+# Backend specific specializations
+def cuda_specializations(specializations, params, args):
+    return specializations
 
+
+def hip_specializations(specializations, params, args):
+    return specializations
```
I don't understand the point of those? It seems to be doing the same thing. Is this where the specific specialization would happen?

giuseros replied:
Yes, this is where I thought the specific specialization would happen. For now the are `nop`.

ThomasRaoux reviewed (commented):
This doesn't seem to handle all the places where we deal with specialization like `compute_spec_key` (this one is a bit sensitive as it is used to compute the hash key and therefore is sensitive to execution time). 
Also I was hoping for something where we call into the existing backend classes. I haven't looked in details at how easy it would be to do though.

giuseros commented:
Hi @ThomasRaoux , I tried to apply your suggestions:
- I moved the class to store the properties of the different parameters into the common backend compiler class
- I added two methods to the `BaseBackend` class: `get_attrs_descriptor` and `compute_spec_key`. These will be used through out the different code generation phases (i.e., to generate the properties and create a key to retrieve those properties
- Now the properties are stored not as tuple anymore. The value of the property (16 or 1, in the current state of things) is stored in a different structure. 

Please take another look and let me know what you think!

antiagainst commented on python/triton/backends/compiler.py:
```diff
@@ -8,6 +9,101 @@
+        attrs = {}
+        for prop_name, arg_set in self.arg_properties.items():
+            for arg in arg_set:
+                attrs[arg] = attrs.get(arg, []) + [(prop_name, 16)]
```
We should query `property_val` instead of hardcoding `16`?

antiagainst commented on python/triton/backends/compiler.py:
```diff
@@ -8,6 +9,101 @@
+    arg_properties: dict[str, list]
+    property_val: dict[str, int]
+
+    def __init__(self, params=None, args=None):
```
Add some docstring about the input?

antiagainst commented on python/triton/backends/compiler.py:
```diff
@@ -8,6 +9,101 @@
 from types import ModuleType
 
 
+# This class handles the properties for the given function parameters
```
Use docstring for this doc

antiagainst commented on python/triton/backends/compiler.py:
```diff
@@ -8,6 +9,101 @@
+                attrs[arg] = attrs.get(arg, []) + [(prop_name, 16)]
+        return attrs
+
+    # Return the same object, without the given attribute `attr_name`
```
Same here and all following methods

antiagainst commented on python/triton/backends/compiler.py:
```diff
@@ -8,6 +9,101 @@
+
+    # Create the class from a set of hints that are passed out. So instead
+    # of calling the specializations, it's the user that tells us what property
+    # each argument has
```
Can we also add doc regarding what structure `hints` should be?

antiagainst commented on python/triton/backends/compiler.py:
```diff
@@ -82,3 +178,17 @@ def get_module_map(self) -> Dict[str, ModuleType]:
+        """
+        Return an attribute descriptor: given a set of parameters and arguments
+        the descriptor stores a set of compile time properties that can improve code
+        generation. Different backends might benefit from different properties
```
End the sentense with `.`. Similarly for the following.

antiagainst commented on python/triton/backends/compiler.py:
```diff
@@ -8,6 +9,101 @@
+    def from_hints(hints):
+        attrsDescriptor = AttrsDescriptor()
+        for prop_name, prop_val in attrsDescriptor.property_val.items():
+            attrsDescriptor.arg_properties[prop_name] = [i for i, h in hints.items() if h == prop_val]
```
Hmm, `hints` should differentiate between divisble by 16 and equal to 1?

giuseros replied:
Why? I mean, `hints` is a list like `[(0, 1), (1,16), (2, 16), (3,1)....]`. So the `AttrsDescriptor` can parse this list and form the different properties.

antiagainst replied:
Okay I guess I was just quite confused about the hints format. The docstring helps to clarify here. Thanks.

antiagainst commented on python/triton/backends/compiler.py:
```diff
@@ -8,6 +9,101 @@
+        return attrsDescriptor
+
+    @staticmethod
+    def is_divisible_by_16(x):
```
Add docstrings for the following methods

antiagainst commented on python/triton/compiler/code_generator.py:
```diff
@@ -1280,17 +1280,18 @@ def ast_to_ttir(fn, specialization, context, options, codegen_fns, module_map):
-    new_attrs = {k: [("tt.divisibility", 16)] for k in attrs.divisible_by_16}
-
+    new_constants = {k: True if k in tys and tys[k] == "i1" else 1 for k in attrs["tt.equal_to_1"]}
+    print(new_constants)
```
`print` debug leftover.

antiagainst commented on python/triton/compiler/code_generator.py:
```diff
@@ -1264,9 +1264,9 @@ def kernel_suffix(signature, specialization):
+        if i in specialization["tt.equal_to_1"]:
             suffix += 'c'
-        if i in specialization.divisible_by_16:
+        if i in specialization["tt.divisibility"]:
```
Do we need to have ways to handle additional properties injected by backends here?

giuseros replied:
So I was a bit unsure about this. This method (apparently) is only used by the AOT driver. I can bake this in `AttrsProperty`, but I wanted to know if that was the right call. Because it looks that `AttrsProperty` is more dedicated for non-AOT/standard  compilation flow. 

Alternatively, we could move this method in `compile.py` and let different backends add their own AOT specialization in there.

antiagainst replied:
Yeah okay fine to leave it as-is for now.

antiagainst commented on python/triton/backends/compiler.py:
```diff
@@ -8,6 +9,101 @@
+        return attrs
+
+    # Return the same object, without the given attribute `attr_name`
+    def erase_property(self, attr_name):
```
`erase_property` implies inplace update to me.. I'd name this as `filter_out_property` here to be clear.

antiagainst commented on python/triton/runtime/jit.py:
```diff
@@ -640,11 +604,11 @@ def run(self, *args, grid, warmup, **kwargs):
                 p.name: v
                 for (v, p) in zip(bound_vals, self.params)
-                if p.is_constexpr or p.num in configs[0].equal_to_1 or v is None
+                if p.is_constexpr or p.num in configs[0]["tt.equal_to_1"] or v is None
```
This is hardcoding some properties. Do you see needs to allow each backend to inject something here? I guess it's hard given this is touching compiler-runtime interface.. Maybe we should be clear in the requirements for backend-specific properties at least.

giuseros replied:
If I understand your comment, you are saying that a backend might not implement a "tt.equal_to_1" property. If this is the case, this will error. So either a) we impose the backend's AttrsDescriptor to have a "tt.equal_to_1" property, or b) we need a way to pass `p` or `p.num` to AttrsDescriptor to let the class tell us if `p` is a constant.

My idea was a bit to go with a). I implemented a `__get_item__` function in the base class that returns None if "tt.equal_to_1" is not there. So if the backend does not implement "tt.equal_to_1" this test will simply return `False` 

 To go with b) direction,I can add a method `is_constant(param)`. The default implementation will be to look a the "tt.equal_to_1" property.

antiagainst replied:
I'm actually asking the case where some additional backend specific properties may want to do something like `equal_to_1` is doing, which requires touching the interface when calling the kernel. I guess we can worry this a bit later when the needs arise and cap the backend-specific properties to not support such cases for now.

antiagainst reviewed (changes requested):
Thanks for this! A few comments inlined. Bigger questions are:

1) There are hardcoded logic to check current supported properties in runtime jit. We should think about the limitations on backend specific properties to be clear on what they can/cannot do and document. It's fine to only enable backend specific properties to just influence code generation alone IMO.
2) I'm wondering whether we can make the `AttrsDescriptor` better structured with `__slots__`.

antiagainst commented on python/triton/backends/compiler.py:
```diff
@@ -8,6 +9,139 @@
+        }
+
+    """
+    __slots__ = ('arg_properties', 'property_val', '__dict__')
```
Actually I'm thinking something like `__slots__ = ('divisibility', 'equal_to_1', 'property_values')`. This way allow we only recognize `divisibility` and `equal_to_1` by default and allow subclasses to add more there. It also allows us to use `divisibility`/`equal_to_1`/etc directly as an attribute. wdyt?

giuseros replied:
Done. I am not sure we need `property_values` as slot, so I left it out.

antiagainst commented on python/triton/backends/compiler.py:
```diff
@@ -8,6 +9,139 @@
+
+    def __init__(self, params=None, values=None):
+        """
+        We can initialize the AttrsDescriptor class by passing the list of params
```
Nit: typically we have a short oneliner to explain concisely and then a line break and then more detailed explanation. Similary for other docstrings.

antiagainst commented on python/triton/backends/compiler.py:
```diff
@@ -8,6 +9,139 @@
+        parameters. I.e., the dictionary will look like:
+        {
+        "prop0": (0, 2, 3)
+        "prop1": (0, 4, 5)
```
Also mention that the mapped to value are parameter indices?

giuseros commented:
I had a go trying to reproduce NVIDIA failures without luck: it seems ok. Let's see if it was a glitch, otherwise I will work harder in reproducing the error

antiagainst commented on python/triton/backends/compiler.py:
```diff
@@ -8,6 +9,166 @@
+        return attrs
+
+    def get_constants(self) -> Dict:
+        """ Return the a dict of constant properties and their values """
```
.. indices of constant parameters to their values ..

antiagainst commented on python/triton/backends/compiler.py:
```diff
@@ -8,6 +9,166 @@
+        - "tt.divisibility", i.e., is the given parameter divisible by 16
+        - "tt.equal_to_1", i.e., is the given parameter an integer constant 1
+
+    `arg_value`: a dictionary containing the value of the different compile-time properties, like:
```
Need to update the name and also mention `constant_properties`

antiagainst reviewed (commented):
Hopefully the last batch of comments. :)

giuseros commented on python/triton/backends/compiler.py:
```diff
@@ -34,21 +35,26 @@ class AttrsDescriptor:
 
     """
-    __slots__ = ('arg_properties', 'property_val', '__dict__')
+    __slots__ = ('divisibility', 'equal_to_1', '__dict__')
```
It was left there because the hash function was using `__dict__.values()`. I reimplemented it, so now I can remove it

giuseros commented on python/triton/backends/compiler.py:
```diff
@@ -8,6 +9,166 @@
+
+    def init_slots(self):
+        """ Initialize the slots of this class """
+        self.divisibility = self.arg_properties["tt.divisibility"]
```
The problem is that the object is created in three different ways:
- By constructor
- `from_dict`
- `from_hint`
They all need to initialize the slots. So I ended up leaving `init_slots` and adding `add_common_properties` and `add_additional_properties` . This last one is supposed to be reimplemented by the backend.

giuseros commented:
Hi @antiagainst , I addressed your comments. PTAL.  I still see and NVIDIA failure, but I want to get the code right before diving into that.

antiagainst commented on python/triton/backends/compiler.py:
```diff
@@ -8,6 +9,184 @@
+        self.add_additional_properties(params, values)
+        self.init_slots()
+
+    def add_common_properties(self, params, values):
```
Typically we prefix `_` for methods that are internal to the class. See https://stackoverflow.com/a/1301369. I think here we want to apply to `_add_common_properties`, `_add_additional_properties`, `_init_slots`.

antiagainst commented on python/triton/backends/compiler.py:
```diff
@@ -8,6 +9,184 @@
+            if AttrsDescriptor.is_equal_to_1(arg) and not param.do_not_specialize
+        ]
+
+    def add_additional_properties(self, params=None, values=None):
```
On a second thought, maybe `_add_backend_properties` is clearer.

antiagainst reviewed (approved):
Nice, LGTM now! Thanks for bearing with my nitpicking. :) I'll leave the final approve to Thomas though.

giuseros commented:
Thank you for your review @antiagainst !

giuseros commented:
FYI, [this is how I intend to use this refactor on the AMD backend](https://github.com/triton-lang/triton/pull/4716/files#diff-bce8d4dc00e2d96055a9cbc98e6eae7aba3b76d70a6409c6788acb3b32755049L20)

ThomasRaoux commented on python/triton/compiler/code_generator.py:
```diff
@@ -1270,7 +1270,7 @@ def kernel_suffix(signature, specialization):
         if i in specialization.equal_to_1:
             suffix += 'c'
-        if i in specialization.divisible_by_16:
+        if i in specialization.divisibility:
             suffix += 'd'
```
I don't understand how it works. Does divisibility means divisible by 16? If it can be other divisibility then would they all have the same suffix?

antiagainst replied:
IIUC we only support `divisibility=16` cases before.

giuseros replied:
Hi @ThomasRaoux , I guess it's a matter of design choice. The original implementation was very specific. The specialization for divisibility was setting always the divisibility to 16 (like @antiagainst  is mentioning). 

So now we have  that `AttrsDescriptor.arg_properties` is a `map(attr_name, param)` and  all the params will be set to a single value in `property_values` (so that each element in `param_set` will have that single value)

We can decide to be more generic and have a `arg_properties` like a `map(attr_name, (param_set, value))` and get rid of the `property_values` map.

ThomasRaoux replied:
my point was more: either this is specific to divisibility16 and it should be explicit in the naming or the suffix should be different based on the divisibility value?

giuseros replied:
Alright, if it is how we represent the attribute from the class, I made the change that now the slots are `divisibility_16` and `equal_to_1`, so it should be clearer what we are asking for

ThomasRaoux reviewed (commented):
I think that's fine, just added one comment. I do wonder about the runtime performance impact. I don't see any redflags but I'll check if this is something we have an easy way to benchmark

giuseros commented:
Hi @ThomasRaoux , @antiagainst , any updates on this?

ThomasRaoux reviewed (approved):
LGTM

alexbaden commented on python/triton/backends/compiler.py:
```diff
@@ -8,6 +9,184 @@
+
+    def _add_common_properties(self, params, values):
+        """ Add common compile-time properties """
+        self.property_values["tt.divisibility"] = 16
```
With this change it seems like it is not possible to specify divisibility of different values for different parameters (e.g. https://github.com/pytorch/pytorch/blob/main/test/inductor/test_metrics.py#L21). Was that intended, or is it just a side effect?

alexbaden replied:
Ok, thanks. PyTorch Inductor has an option to apply a `divisibility=8` attribute but it is not clear that option is used - it is only set if `AttrsDescriptor` has that property and I have not seen any evidence of that property in the codebase stretching back to the big refactor last November.
